lib/batch_loader.rb

```diff
@@ -23,11 +23,18 @@ def initialize(item:, executor_proxy: nil)
     @__executor_proxy = executor_proxy
   end
 
-  def batch(default_value: nil, cache: true, key: nil, &batch_block)
+  def batch(default_value: nil, cache: true, replace_methods: nil, key: nil, &batch_block)
     @default_value = default_value
     @cache = cache
+    # Default to the value of `cache` unless explicitly set
+    @replace_methods = replace_methods.nil? ? @cache : replace_methods
     @key = key
     @batch_block = batch_block
+
+    if @replace_methods && !@cache
+      raise ArgumentError, "Can't replace methods without caching the value"
```

Comment:
what do you think about simplifying the code and removing this check? could there be any unexpected effects?

e.g. `replace_methods: true, cache: true` by default (keeping the compatibility) and they don't depend on each other

Reply:
I'm open to it, but I'm not sure what `cache: false, replace_methods: true` would actually mean. If you set `cache: false` then you want to retrieve the value each time, but if you set `replace_methods`, you're running a bunch of `#define_method`s to point to a _specific_ value that was retrieved at one point.

If you think that's not a problem I'm happy to simplify though :smiley:

Reply:
yeah, could you please simplify and "decouple" the arguments? 🙏

that will simplify the code. ideally `cache: false` should be used rarely. but I could imagine a case when devs would prefer using a custom cache (e.g. across multiple nodes) and specify `replace_methods: true` to make the lookups and usage of the object faster later.

the rest of the PR looks excellent!

Reply:
OK, I made that change! I did have to change two specs because the message expectations relied on `replace_methods: false` (implicitly), so I also added a note to the README that you probably want `replace_methods: false` with `cache: false`.